apm-agent-plugins/apm-scala-concurrent-plugin/src/main/scala/co/elastic/apm/agent/concurrent/FutureInstrumentation.scala

```diff
@@ -0,0 +1,41 @@
+package co.elastic.apm.agent.concurrent
+
+import java.util
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation
+import net.bytebuddy.asm.Advice
+import net.bytebuddy.description.`type`.TypeDescription
+import net.bytebuddy.description.method.MethodDescription
+import net.bytebuddy.matcher.ElementMatcher
+import net.bytebuddy.matcher.ElementMatchers._
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+class FutureInstrumentation extends ElasticApmInstrumentation {
+
+  override def getTypeMatcher: ElementMatcher[_ >: TypeDescription] =
+    hasSuperType[TypeDescription](named("scala.concurrent.Future"))
+      .or(hasSuperType(named("scala.concurrent.impl.Promise")))
+      .or(hasSuperType(named("scala.concurrent.impl.Promise$Transformation")))
+      .or(hasSuperType(named("scala.concurrent.Future$")))
+
+  override def getMethodMatcher: ElementMatcher[_ >: MethodDescription] =
+    named[MethodDescription]("onComplete").and(returns(classOf[Unit])).and(takesArguments(classOf[Try[_] => _]))
+//    .or(named[MethodDescription]("transform").and(returns(classOf[Future[_]])).and(takesArguments(classOf[Try[_] => Try[_]])))
+//    .or(named[MethodDescription]("transformWith").and(returns(classOf[Future[_]])).and(takesArguments(classOf[Try[_] => Future[_]])))
+    .and(not(isTypeInitializer[MethodDescription]()))
+
+  override def getInstrumentationGroupNames: util.Collection[String] = List("concurrent", "future").asJavaCollection
+
+  @Advice.OnMethodEnter(suppress = classOf[Throwable])
+  def onComplete(@Advice.Argument(value = 0, readOnly = true) callback: Try[_] => _): Unit = {
```

Comment:
What are you trying to achieve by instrumenting the `onComplete` method? 

Is the `onComplete` executed in the context of the `Runnable` from the `Executor` that is used within the `ExecutionContext.fromExecutor`?

Reply:
What should happen is when a `Future` is created, it should have the trace information. When a `Future` finishes its task, the method `onComplete` will be called on that same `Future`. Based on the trace information on the creation, it should then close the correct spans.

Reply:
Another remark on using Java to write the Scala instrumentation. For example, this method would need to be replaced with some custom code to correctly close tracing contexts.
When writing all of this in Java, I don't immediately see a way to write the correct Scala code to replace the original method?